File: day06/day06.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	content, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read input.txt: %v\n", err)
		os.Exit(1)
	}

	// Split and filter empty lines
	allLines := strings.Split(string(content), "\n")
	var lines []string
	for _, line := range allLines {
		if line != "" {
			lines = append(lines, line)
		}
	}
	h := len(lines)

	// Find max width
	w := 0
	for _, line := range lines {
		if len(line) > w {
			w = len(line)
		}
	}

	// Pad lines to same width
	grid := make([][]rune, h)
	for i, line := range lines {
		row := make([]rune, w)
		for j := 0; j < w; j++ {
			if j < len(line) {
				row[j] = rune(line[j])
			} else {
				row[j] = ' '
			}
		}
		grid[i] = row
	}

	// Get a column as runes
	getCol := func(c int) []rune {
		col := make([]rune, h)
		for r := 0; r < h; r++ {
			col[r] = grid[r][c]
		}
		return col
	}

	// Check if column is all spaces (separator)
	isSep := func(c int) bool {
		col := getCol(c)
		for _, ch := range col {
			if ch != ' ' {
				return false
			}
		}
		return true
	}

	// Find problem column groups
	var problems [][]int
	var current []int

	for c := 0; c < w; c++ {
		if isSep(c) {
			if len(current) > 0 {
				problems = append(problems, current)
				current = nil
			}
		} else {
			current = append(current, c)
		}
	}
	if len(current) > 0 {
		problems = append(problems, current)
	}

	// Solve a problem
	solve := func(cols []int, vertical bool) int64 {
		// Find operator
		var op func(int64, int64) int64
		for _, c := range cols {
			ch := grid[h-1][c]
			if ch == '+' {
				op = func(a, b int64) int64 { return a + b }
				break
			} else if ch == '*' {
				op = func(a, b int64) int64 { return a * b }
				break
			}
		}
		if op == nil {
			return 0
		}

		var nums []int64

		if vertical {
			// Each column is a number read top-to-bottom
			for _, c := range cols {
				var sb strings.Builder
				for r := 0; r < h-1; r++ {
					ch := grid[r][c]
					if ch >= '0' && ch <= '9' {
						sb.WriteRune(ch)
					}
				}
				s := sb.String()
				if s != "" {
					n, err := strconv.ParseInt(s, 10, 64)
					if err == nil {
						nums = append(nums, n)
					}
				}
			}
		} else {
			// Each row is a number read left-to-right
			for r := 0; r < h-1; r++ {
				var sb strings.Builder
				for _, c := range cols {
					sb.WriteRune(grid[r][c])
				}
				s := strings.TrimSpace(sb.String())
				if s != "" {
					n, err := strconv.ParseInt(s, 10, 64)
					if err == nil {
						nums = append(nums, n)
					}
				}
			}
		}

		if len(nums) == 0 {
			return 0
		}

		result := nums[0]
		for i := 1; i < len(nums); i++ {
			result = op(result, nums[i])
		}
		return result
	}

	// Part 1: sum of all problems solved horizontally (left-to-right)
	var part1 int64
	for _, p := range problems {
		part1 += solve(p, false)
	}

	// Part 2: sum of all problems solved vertically (top-to-bottom), in reverse order
	var part2 int64
	for i := len(problems) - 1; i >= 0; i-- {
		part2 += solve(problems[i], true)
	}

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}

File: day06/day06.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input.txt").expect("Failed to read input.txt");
    let lines: Vec<&str> = input.lines().collect();
    let h = lines.len();
    let w = lines.iter().map(|l| l.len()).max().unwrap_or(0);

    // Pad lines to same width
    let grid: Vec<Vec<char>> = lines
        .iter()
        .map(|l| {
            let mut row: Vec<char> = l.chars().collect();
            row.resize(w, ' ');
            row
        })
        .collect();

    // Get a column as chars
    let col = |c: usize| -> Vec<char> { (0..h).map(|r| grid[r][c]).collect() };

    // Check if column is all spaces (separator)
    let is_sep = |c: usize| col(c).iter().all(|&ch| ch == ' ');

    // Find problem column groups
    let mut problems: Vec<Vec<usize>> = Vec::new();
    let mut current: Vec<usize> = Vec::new();

    for c in 0..w {
        if is_sep(c) {
            if !current.is_empty() {
                problems.push(current.clone());
                current.clear();
            }
        } else {
            current.push(c);
        }
    }
    if !current.is_empty() {
        problems.push(current);
    }

    // Solve a problem
    let solve = |cols: &[usize], vertical: bool| -> i64 {
        // Find operator
        let op: fn(i64, i64) -> i64 = cols
            .iter()
            .find_map(|&c| match grid[h - 1][c] {
                '+' => Some((|a, b| a + b) as fn(i64, i64) -> i64),
                '*' => Some((|a, b| a * b) as fn(i64, i64) -> i64),
                _ => None,
            })
            .expect("No operator found");

        let nums: Vec<i64> = if vertical {
            // Each column is a number read top-to-bottom
            cols.iter()
                .filter_map(|&c| {
                    let s: String = (0..h - 1)
                        .map(|r| grid[r][c])
                        .filter(|ch| ch.is_ascii_digit())
                        .collect();
                    if s.is_empty() {
                        None
                    } else {
                        s.parse().ok()
                    }
                })
                .collect()
        } else {
            // Each row is a number read left-to-right
            (0..h - 1)
                .filter_map(|r| {
                    let s: String = cols.iter().map(|&c| grid[r][c]).collect::<String>();
                    let trimmed = s.trim();
                    if trimmed.is_empty() {
                        None
                    } else {
                        trimmed.parse().ok()
                    }
                })
                .collect()
        };

        nums.into_iter().reduce(op).unwrap_or(0)
    };

    let part1: i64 = problems.iter().map(|p| solve(p, false)).sum();
    let part2: i64 = problems.iter().rev().map(|p| solve(p, true)).sum();

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);
}
